### PR TITLE
prepare for GUI node selection; prefer global node if newer

### PR DESCRIPTION
**Pull request by luclu**
See https://github.com/ethereum/mist/pull/2093

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -621,22 +622,23 @@ class Manager {
...
     })
     .then((output) => {
-      const haystack = output.stdout + output.stderr;
+      this._logger.debug(`Sanity check output: ${output.stdout + output.stderr}`);
 
-      this._logger.debug(`Sanity check output: ${haystack}`);
-
-      const needles = sanityCheck.output || [];
+      try {
+        var version = output.stdout.match(/\d+\.\d+\.\d+/)[0];
```
Comment (hiddentao) [inline src/index.js:628]: We should not be hardcoding version checks. The `sanityCheck.output` configuration key is meant to contain the output that needs to checked for.

File: src/index.js [hunk 2]
```diff
@@ -621,22 +622,23 @@ class Manager {
...
+      } catch (err) {
+        throw new Error(`Unable to find semantic version >= "${client.version}" in "${binPath} ${sanityCheck.args}"'s' output`);
+      }
 
-      for (let needle of needles) {
-        if (0 > haystack.indexOf(needle)) {
-          throw new Error(`Unable to find "${needle}" in ${client.id} output`);
-        }
+      if (version && semver.lt(version, client.version)) {
```
Comment (hiddentao) [inline src/index.js:633]: We should not be hardcoding this semver check. We should instead refactor the sanity check configuration system such that such logic can be easily represented. I would recommend adding a `versionCommand` config parameter which specifies the CLI arguments to use in order to obtain binary version. Then we can do this semver check.

**Review by hiddentao — CHANGES_REQUESTED** [review]
Good ideas, but we mustn't hardcode this logic. Need to refactor to keep things flexible. Let me know if you'd me to have a go at doing this.

File: src/index.js (hunk 2 above)
Comment (luclu) [inline src/index.js]: You are right, that would be better. 

Though, I can only think of using `stdout.match(/\d+\.\d+\.\d+/)[0]` for the version string hoping that every `<node> [--]version` will have its version as the first semantic version string in the output. 

Eventually the haystack-method is fine, but then a method to retrieve the system node's version via a regex in the `clientBinaries.json` might bring a bit more flexibility.

Alternatively we could omit the version checking and supply an optional boolean `--uselocalnode` quite similar to `--gethpath`.

What do you think?

**Comment by luclu** [comment]
You are right, that would be better. 

Though, I can only think of using `stdout.match(/\d+\.\d+\.\d+/)[0]` for the version string hoping that every `<node> [--]version` will have its version as the first semantic version string in the output. 

Eventually the haystack-method is fine, but then a method to retrieve the system node's version via a regex in the `clientBinaries.json` might bring a bit more flexibility.

Alternatively we could omit the version checking and supply an optional boolean `--uselocalnode` quite similar to `--gethpath`.

What do you think?

**Comment by hiddentao** [comment]
I think we could have a built-in type of sanity check that does the version check you require, so. e.g:

```js
{
    "clients": {
        "Geth": {
            "version": "1.5.5",
            "platforms": {
                "linux": {
                    "x64": {
                        "download": {
                            "url": "https://gethstore.blob.core.windows.net/builds/geth-linux-amd64-1.5.5-ff07d548.tar.gz",
                            "type": "tar",
                            "sha256": "4f36d6df25c37eb33829407d8bf2cea209cc412b3443319e2430db18581c7c01",
                            "bin": "geth-linux-amd64-1.5.5-ff07d548/geth"
                        },
                        "bin": "geth",
                        "commands": {
                            /* Mandatory version check - ensuring binary is >= expected version */
                            "version": {
                                "args": "version",
                                "extract": <regex to extract version>
                            }
                            /* Optional sanity check to ensure binary is what we expect */
                            "sanity": {
                                "args": "some command",
                                "needle": <regex to search for>
                            }
                        }
                    }
                }
            }
        }
    }
}
```

Thus it would call `geth version` and then check that the version output by the binary is >= to the version specified earlier on in the config. We could still keep the current sanity check system for additional checks.

**Pull request closed without merging**